internal/inputs/generator_queries.go

```diff
@@ -0,0 +1,272 @@
+package inputs
+
+import (
+	"bufio"
+	"encoding/gob"
+	"flag"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/databases/cassandra"
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/databases/clickhouse"
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/databases/influx"
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/databases/mongo"
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/databases/siridb"
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/databases/timescaledb"
+	"github.com/timescale/tsbs/cmd/tsbs_generate_queries/utils"
+)
+
+// Error messages when using a QueryGenerator
+const (
+	ErrInvalidQueryConfig = "invalid config: QueryGenerator needs a QueryGeneratorConfig"
+	ErrEmptyQueryType     = "query type cannot be empty"
+
+	errBadQueryTypeFmt        = "invalid query type for use case '%s': '%s'"
+	errCouldNotDebugFmt       = "could not write debug output: %v"
+	errCouldNotEncodeQueryFmt = "could not encode query: %v"
+	errCouldNotQueryStatsFmt  = "could not output query stats: %v"
+)
+
+// QueryGeneratorConfig is the GeneratorConfig that should be used with a
+// QueryGenerator. It includes all the fields from a BaseConfig, as well as
+// options that are specific to generating the queries to test against a
+// database, such as the query type and individual database options.
+type QueryGeneratorConfig struct {
+	BaseConfig
+	QueryType            string
+	InterleavedGroupID   uint
+	InterleavedNumGroups uint
+
+	// TODO - I think this needs some rethinking, but a simple, elegant solution escapes me right now
```

Comment:
Hm, yeah, I'm drawing blanks on a clever solution here as well. But I do see this becoming a pain point as we add more DBs and additional functionality for queries that get generated. One possibility is to have a single `QueryConfig` field that is defined as an `interface{}` with functions that type infer it to a specific database's configuration when needed for validation or generation, but I suppose we'll cross that bridge when we get there.